weave/trace/ipython.py

```diff
@@ -0,0 +1,55 @@
+"""Utilities for working with IPython/Jupyter notebooks."""
+
+import ast
+from typing import Callable
+
+
+class NotInteractiveEnvironmentError(Exception): ...
+
+
+def is_running_interactively() -> bool:
+    """Check if the code is running in an interactive environment."""
+    try:
+        from IPython import get_ipython
+
+        return get_ipython() is not None
+    except ModuleNotFoundError:
+        return False
+
+
+def get_notebook_source() -> str:
+    """Get the source code of the running notebook."""
+    from IPython import get_ipython
+
+    shell = get_ipython()
+    if shell is None:
+        raise NotInteractiveEnvironmentError
+
+    if not hasattr(shell, "user_ns"):
+        raise AttributeError("Cannot access user namespace")
+
+    # This is the list of input cells in the notebook
+    in_list = shell.user_ns["In"]
+
+    # Stitch them back into a single "file"
+    full_source = "\n\n".join(cell for cell in in_list[1:] if cell)
+
+    return full_source
+
+
+def get_class_source(cls: Callable) -> str:
+    """Get the latest source definition of a class in the notebook."""
+    notebook_source = get_notebook_source()
+    tree = ast.parse(notebook_source)
+    class_name = cls.__name__
+
+    # We need to walk the entire tree and get the last one since that's the most version of the cls
```

Comment:
This could certainly be incorrect if there is more than one definition of the same class name. We should have some token in the code comment that we can use to indicate this like "WARNING:"

Reply:
I think this is safe because this inspection is happening at publish time.  The flow is roughly:
```py
class A(weave.Object):
  a: str
class A(weave.Object):
  b: str
a = A()
weave.publish(a)  # This will publish A:v0 with b: str which is what I would expect
```

This is also expected:
```py
class A(weave.Object):
  a: str
a = A()
weave.publish(a)  # Publishes A:v0 with a: str

class A(weave.Object):
  b: str
a = A()
weave.publish(a)  # Publishes A:v1 with b: str
```

Reply:
I think this is mostly safe because this inspection is happening at publish time.  The flow is roughly:
```py
[In:1]
class A(weave.Object):
  a: str

[In:2]
class A(weave.Object):
  b: str
a = A()
weave.publish(a)  # This will publish A:v0 with b: str which is what I would expect
```

This is also expected:
```py
[In: 1]
class A(weave.Object):
  a: str
a = A()
weave.publish(a)  # Publishes A:v0 with a: str

[In: 2]
class A(weave.Object):
  b: str
a = A()
weave.publish(a)  # Publishes A:v1 with b: str
```

There may be this gotcha:
```py
[In: 1]
class A(weave.Object):
  a: str
a = A()

[In: 2]
weave.publish(a)  # Maybe Publishes A:v0 with b: str

class A(weave.Object):
  b: str
a = A()
weave.publish(a)  # No publish since duplicate?
```